### DISCUSSION TITLE
Buttons to send serial commands back to devices, without entering commands manually 1 by 1.

### DISCUSSION THREAD

**Discussion by Benjamincreator** [Serial-Studio/Serial-Studio#231] [Ideas]
First off, I want to say that Serial Studio is a fantastic app, thank you, and the contributors for creating it. It’s an incredibly powerful tool, and I appreciate the effort that has gone into its development.

From using it a bit, I was wondering: would it be possible to introduce interactive buttons as widgets within Serial Studio, seamlessly integrated alongside existing data visualisation elements?

These buttons could serve dual purposes:
- Sending predefined commands directly to the connected device.
- Triggering dynamic actions based on user-defined parameters or real-time data.

This feature would allow users to control their embedded applications while simultaneously visualising the resulting changes in system behaviour.

Example Use Cases:
- Device Control: A button could send a command to enable a device (e.g., CMD_ENABLE) and update its visual state based on received serial feedback. For instance, the button could turn red when the device is off and green when it’s on. Depending on the incoming data (e.g., device_status:1,fan_speed:60,current_brightness:128).
- Dynamic Response: Buttons could act as manual toggles or respond automatically to specific conditions in the data stream. (e.g., temperature exceeding a threshold, low system voltage, or other events). For example, in a model rocketry setup, a button could display the continuity status of an ignitor. If the continuity is bad, and the system sends this data, the button could visually reflect the issue.

Additional Features:
- Custom Button Layouts: Users could arrange buttons flexibly alongside graphs, gauges, and other data displays to suit their workflow.
- Configurable Inputs: Widgets like sliders, dropdown menus, or other inputs could be used to send commands or adjust settings (e.g., brightness levels or operational modes).
- Dynamic Feedback Integration: Buttons and widgets would update in real-time based on parsed serial data, reflecting the state of the system (e.g., toggling colours or states based on feedback).

This functionality could expand Serial Studio’s capabilities significantly, turning it into a powerful control and monitoring interface for embedded systems. Thank you for considering this idea, and again, thank you for the fantastic work on Serial Studio!

Cheers Ben!

**Comment by alex-spataru** [comment]
Hi, this feature is already supported with the [Actions](https://github.com/Serial-Studio/Serial-Studio/wiki/Project-Editor#action-view-when-an-action-is-selected) in the Project Editor. For the moment, only buttons with constant data input are supported, but maybe in the future I'll add support for more widgets.

You can see more info in the original issue/proposal here: https://github.com/Serial-Studio/Serial-Studio/issues/205

**Reply by Mark-Oberst** [reply]
Awesome! I just found Serial Studio and I really like it – fantastic work! The ease of getting started makes it especially useful. 

I wanted to chime in on this discussion because I think a dynamic action button would be incredibly useful for many use cases. For example, I’m working on a stepper motor-based test system, and I would love to be able to adjust parameters like motor current, acceleration, and velocity directly through Serial Studio, rather than hardcoding them into my microcontroller. It would be awesome if that feature were implemented at some point. Anyways, thank you for the work so far. :)
